### PR TITLE
[Erndata] Support Erndata MTP magic send

### PR DESCRIPTION
**Pull request by xuxinyi389**
### PR Category

Distributed Strategy

### PR Types

New features, Bug fixes

### Description

Enable `use_erndata=True` with MTP `enable_mtp_magic_send=True` under pipeline parallelism.

Why:
- Erndata carries length-L packed sequences and `cu_seqlens_q`, while the legacy magic-send path assumes the historical L+K contract and process-global input-id state.
- PP stages need an explicit, dynamic-shape metadata transport and synchronized physical copies of the MTP token embedding.

What:
- Carry `mtp_full_input_ids` and `cu_seqlens_q` explicitly through the pipeline while keeping the backbone carrier at 1x.
- Re-embed and packed-roll decoder inputs on each MTP stage, then apply the configured CP/SP layout.
- Synchronize MTP embedding initialization and gradients across PP stages, including symmetric zero-gradient participation.
- Reset dynamic P2P metadata caches and validate unsupported embedding/dropout combinations.
- Add unit, PP2, K>=2 mHC, and PP2/CP2 multi-step embedding-consistency coverage.

### Does this PR introduce any accuracy change?

Yes, intentionally fixes correctness for Erndata MTP magic-send. The magic and non-magic packed decoder embeddings are asserted equal for K=1/K=2, packed document tails remain zero, and the PP2/CP2 test verifies stage-0 and MTP embedding parameters remain bit-identical across ten optimizer steps.

### Verification

- Focused unit suite: 135 passed, 7 subtests passed.
- Ruff check and Ruff format check pass on all changed Python files.
- Four-GPU PP2 x CP2 synchronization test: all four ranks pass.
- Prior PP2 two-GPU smoke and 16-GPU ten-step Erndata acceptance run reached finite MTP loss; the separate rank-0 process teardown issue occurs after training completion.

### CODE REVIEW

**Comment by risemeup1111** [comment]
### Paddle-Bot Review Board （review完成)

| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4038103975">Codestyle 未过：4 个文件需 ruff&#45;format，新增用例另有 ruff&#45;check F401</a> | P2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;代码质量</a> | ✅ |
| 2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4036954589">共享权重梯度同步的静态图分支缺少 None 兜底</a> | P3 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;功能正确性与兼容性</a> | ✅ |
| 3 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4036953049">mtp&#95;embed 广播缓冲区大小用 assert 做运行时校验</a> | P3 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;安全与错误处理</a> | ✅ |
| 4 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4036956213">magic send 与残差多流（K≥2）组合缺少测试覆盖</a> | P2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;测试质量</a> | ✅ |
| 5 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#pullrequestreview-5235782273">PR 描述模板字段仍未填写，Check PR Description 失败</a> | P3 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;PR 信息与评论</a> | ✅ |
| 6 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4037167183">新增的 CP 缩放回归用例断言对象被测试替换，拦不住回归</a> | P2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;测试质量</a> | ✅ |
| 7 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2088#discussion_r4038102447">新增 CP2 用例未登记 4 卡，CI 以 2 卡启动拿不到覆盖</a> | P2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/blob/5cc8cc3559331cb902c3b3bea798161ced2bacbe/.agents/skills/ai-review/references/base-rules.md">仓库规则：基础规则&#45;测试质量</a> | ✅ |


<sub>Powered by Nyanpasu claude with Opus 4&#46;8 默认推理级别&#44; please check the suggestions carefully&#46;</sub>

File: tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py [hunk 1]
```diff
@@ -157,6 +162,25 @@ def run_pp(seed, batch_size, seq_len, vocab_size, cu_seqlens_list):
...
+    # Magic send must materialize a local MTP embedding on whichever PP rank
+    # owns the MTP block.  Aggregate the fact across ranks so this assertion is
+    # meaningful on every process, including stage 0.
+    local_mtp_layers = gpt_model._get_all_mtp_layers()
+    local_has_mtp_embed = any(
+        layer.mtp_embed is not None for layer in local_mtp_layers
+    )
+    has_mtp_embed = paddle.to_tensor(
+        [int(local_has_mtp_embed)], dtype="int32"
+    )
```
Comment (risemeup1111) [inline tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py]: P2
本文件与 `tests/single_card_tests/transformer/test_erndata_mtp_granularity_and_mhc_chain.py` 未按仓库格式提交，导致必过的 Codestyle Check 在 `ruff-format` 钩子上失败（CI 日志：`2 files reformatted`）。对本次改动的 14 个 Python 文件执行 `ruff format --check`，报出需要重排的正是这两个文件，例如本处 `paddle.to_tensor([...], dtype="int32")` 可以折叠为一行。建议运行 `pre-commit run --all-files`（或 `ruff format`）后把格式化结果一并提交。

File: src/paddlefleet/models/gpt/gpt_model.py [hunk 2]
```diff
@@ -1114,18 +1114,14 @@ def _synchronize_mtp_embed_weight(self):
...
-            padded_vocab = int(
-                (
-                    (self.config.vocab_size + (divisor * tp) - 1)
-                    // (divisor * tp)
-                )
-                * (divisor * tp)
+            assert self.config.vocab_size % tp == 0, (
+                f"vocab_size={self.config.vocab_size} must be divisible by "
+                f"tensor_model_parallel_size={tp}"
             )
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/gpt_model.py]: P3
这里用 `assert` 承担运行时的形状校验：`python -O` 会移除断言，此时 `local_vocab = vocab_size // tp` 在 `vocab_size % tp != 0` 时会静默取到错误大小（`VocabParallelEmbedding` 走 `divide` 取 ceil，非整除时每卡的 `num_embeddings_per_partition` 并非整除结果），接收端参与广播的缓冲区形状就会与源张量不一致。仓库基础规则要求不用 assert 做运行时输入校验，建议改为 `raise ValueError(...)` 并保留现有提示信息。

File: src/paddlefleet/models/gpt/gpt_model.py [hunk 3]
```diff
@@ -1176,57 +1172,67 @@ def allreduce_shared_weight_gradients(self):
         """
         import paddle
         import paddle.distributed
+        from paddle import framework
+        from paddle.framework import core
+
+        def _attached_grad(param):
+            if not framework.in_dynamic_mode():
+                return param._grad_ivar()
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/gpt_model.py]: P3
非动态图分支直接返回 `param._grad_ivar()`，没有兜底。旧实现在 `grad is None` 时 `continue` 跳过，而新代码在下方共享层循环和 `mtp_embed` 分支都无条件使用返回值，静态图下取到 None 就会以 `NoneType has no attribute 'contiguous'` 崩溃（动态图分支已用零梯度实现“无论如何都参与集合通信”）。建议静态图分支同样补零梯度或显式抛错，保持两个分支语义一致。

File: src/paddlefleet/transformer/multi_token_prediction.py [hunk 4]
```diff
@@ -2085,10 +2141,16 @@ def _forward_megatron_style(self, dict_args: dict) -> dict:
             if self.layer_number < num_nextn - 1:
                 dict_args["mhc_multistream"] = paddle.concat(mhc_chunks)
 
-        # Store back into the pipeline concat tensor at slot k+1.
-        tensor_list[self.layer_number + 1] = hidden_states
+        # Store this depth in the growing carrier. Non-magic mode has a fixed
+        # K+1 carrier and overwrites its precomputed embedding slot instead.
+        if magic_send:
+            tensor_list.append(hidden_states)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_token_prediction.py:2152]: P2
magic send 与残差多流是两条并行通道：本处“载体按 depth 追加”与上方 `mhc_chunks[self.layer_number + 1] = ...` 会同时生效，而 mHC + erndata + magic send（K≥2）是本次新放开的组合，仍无测试覆盖——新增的 mHC 链式用例 `_make_mhc_layer` 均设 `enable_mtp_magic_send=False`，多卡 PP2 用例也未开启残差多流。建议补一个 K≥2 的单卡 magic send + mHC 用例，断言 depth k+1 消费 depth k 的多流输出、且 carrier 各槽宽度一致。

**Review by risemeup1111 — COMMENTED** [review]
P3
PR 描述仍是空白模板（`PR Category`、`PR Types`、`是否引起精度变化` 均未填写），Check PR Description 因此失败；仓库规则要求描述说明为什么改、解决什么问题以及验证方式。本次改动改变了 MTP 分组张量在流水线中的搬运方式与 mtp_embed 的梯度同步，请补齐这些字段，并给出对应的验证结论与精度影响评估。

<div align="right">
   <sup>Powered by Nyanpasu with DeepSeek-V4.1-Flash high, please check the suggestions carefully.</sup>
</div>

File: tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py (hunk 1 above)
Comment (risemeup1111) [reply tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py]: 新 head `14331027` 仍未修复，并且新提交多带进一处：本地对本次改动的 Python 文件执行 `ruff format --check` 现在报 3 个文件需要重排，新 head 的 Codestyle Check 依旧是 `ruff-format` 失败（`3 files reformatted`），多出来的是本提交改动过的 `src/paddlefleet/transformer/multi_token_prediction.py`（`if (...)` 条件的折行写法）。建议推送前先对改动文件执行 `ruff format` / `pre-commit run --all-files`。

File: tests/single_card_tests/transformer/test_mtp_magic_send.py [hunk 5]
```diff
@@ -508,6 +508,29 @@ def test_magic_send_vs_non_magic_send(self):
...
+                variable_seq_lengths=True,
+                context_parallel_size=2,
+            )
+        )
+        self.assertFalse(
+            getattr(
+                layer.mtp_embed.weight,
+                "context_parallel_disable_scale_grad",
+                False,
+            )
```
Comment (risemeup1111) [inline tests/single_card_tests/transformer/test_mtp_magic_send.py:532]: P2
这个回归用例断言的对象不是被标记的那个权重：`_build_mtp_layer` 在构造完 `MultiTokenPredictionLayer` 之后会无条件把 `layer.mtp_embed` 替换成测试自己新建的 `nn.Embedding`，而 `__init__` 里的 `mark_context_parallel_parameter_disable_scale_grad(self.mtp_embed)` 标记的是原 `VocabParallelEmbedding` 的 weight 实例，属性不会迁移到新对象上（实测新 embedding 的 weight 读取该属性恒为 False）。因此即使把新加的 `and not getattr(config, "use_erndata", False)` 去掉，本用例仍会通过，拦不住这次修复要防的回归。建议改成断言构造期真正的那个参数，例如在替换前保存 `layer.mtp_embed.weight` 再对其断言，或直接断言原始 `VocabParallelEmbedding` 的 weight。

File: tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_magic_cp_weight_sync.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            "pp",
+            "sep",
+            "cp",
+            "dp",
+            "ep",
+            "mp",
+        ],
+    }
+    fleet.init(is_collective=True, strategy=strategy)
+    if dist.get_world_size() != WORLD_SIZE:
```
Comment (risemeup1111) [inline tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_magic_cp_weight_sync.py:82]: P2
本用例硬性要求 PP2×CP2=4 rank，但没有在 `tests/test_configs.yaml` 中登记，`ci/multi-card_test.sh` 会命中兜底规则（`tests/multi_card_tests/pipeline_parallel/*.py` → `num_gpus: 2`）以 2 卡启动：此处会因 2≠4 直接 SkipTest，本次新增的回归覆盖在 CI 中等同于不存在。另外 `fleet.init` 排在 world size 判断之前，2 卡下会按 4 卡拓扑建组（EP 路径的 expert 组形如 [0,2]、[1,3]，其中 rank 2/3 并不存在），建组阶段可能阻塞或报错，反而拖住 multi-card 作业。建议参照 `test_gpt_pp_mtp_megatron.py`（2 卡）与 `test_separate_mtp_headloss_pp.py`（8 卡）的登记方式，为本文件补一条 `num_gpus: 4`，并把 world size 检查前移到 `fleet.init` 之前。

File: tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py (hunk 1 above)
Comment (risemeup1111) [reply tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py]: 当前 head 的 Codestyle 仍未通过，共 4 个文件：新增的 `tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_magic_cp_weight_sync.py` 需要 ruff-format 改写第 137、191 行，并有 ruff-check F401（第 42 行 `import paddlefleet` 未使用）；其余为 `src/paddlefleet/transformer/multi_token_prediction.py`、`tests/multi_card_tests/pipeline_parallel/test_gpt_pp_mtp_megatron.py`、`tests/single_card_tests/transformer/test_erndata_mtp_granularity_and_mhc_chain.py`。请在提交前跑一遍 `ruff check --fix` 和 `ruff format`。

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by Waynezee — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM
